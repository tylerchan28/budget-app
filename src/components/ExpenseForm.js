import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import 'react-dates/initialize';
import "react-dates/lib/css/_datepicker.css";

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ""
        } // if props.expense exists, populate input values w/ them
            // if not, set to these values
    
    }

    onDescriptionChange = (e) => { // changes this.state.description whenever value is updated
        const description = e.target.value;
        this.setState(() => ({ description }))
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        } // !amount allows for deletion of amount, now empty value can be set
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        } //if no createdAt, do nothing
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "Please provide description and amount" }))
        } else {
            this.setState(() => ({ error: "" }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            }) // this is the onSubmit function from AddExpensePage
        } // form onSubmit handles the error, props.onsubmit passes in parameters to add an expense

    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}t</p>}
                <form onSubmit={this.onSubmit}> 
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description} // creates read only input, so make onChange
                        onChange={this.onDescriptionChange} // without onChange, cannot write text into input
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

// autoFocus automatically puts cursor on form/can type right when get on page