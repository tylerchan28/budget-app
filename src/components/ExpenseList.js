import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => ( // THIS FUNCTION IS WHAT DISPLAYS THE EXPENSES TO SCREEN
    // takes in entire expenses list (as more expenses are added) and transforms to ExpenseListItem
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div>
                        <span className="list-item list-item--message">No expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
            }
        </div>
    </div>
);
// {...expenses} takes k/v from expense and passes it as props

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);
// API ruires to call connect()(component)
// connects component to Redux store
// as store changes, values update in component (props.filters.text, props.expenses.length)
