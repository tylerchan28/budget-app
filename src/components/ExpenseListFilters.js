import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
        //on each keystroke, changes text filter to input value
    }
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate();
        } else if (e.target.value === "amount") {
            this.props.sortByAmount();
        }
    }
    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} // controlled input, controlled by javascript
                    onChange={this.onTextChange}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId="start"
                    endDate={this.props.filters.endDate}
                    endDateId="end"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={(() => false)}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
// allows user to write to the store through props.dispatch
// connect function allows us to use prop.dispatch rather than store.dispatch
    // react-redux injects dispatch method into props

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
// variables here are same as in their action generators
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);