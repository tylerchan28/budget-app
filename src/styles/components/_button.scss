.button {
    background-color: $blue;
    border: none;
    color: white;
    display: inline-block; // allows for margin/padding to take effect
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1;
    padding: $s-size;
    text-decoration: none;
}

.button--link { // keeps same values but can override
    background: none;
}

.button--secondary {
    background-color: rgb(228, 100, 100)
}