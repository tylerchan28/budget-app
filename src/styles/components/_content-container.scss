.content-container {
    margin: 0 auto; // redistribute left and right space evenly
    max-width: 80rem;
    padding: 0 $m-size;
}