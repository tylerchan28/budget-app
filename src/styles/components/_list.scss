.list-header {
    background-color: $light-grey;
    border: 1px solid darken($light-grey, 8%);
    color: $grey;
    display: flex;
    justify-content: space-between;
    padding: $s-size $m-size;
}

.list-item {
    display: flex;
    border: 1px solid darken($light-grey, 8%);
    border-top: none;
    color: $dark-grey;
    flex-direction: column;
    padding: $s-size;
    text-decoration: none;
    transition: background-color .3s ease;
    &:hover { // shorthand to add hover
        background: $light-grey;
    }
    @media (min-width: $desktop-breakpoint) {
        align-items: center;
        flex-direction: row;
        justify-content: space-between;
        padding: $m-size;
    }
}

.list-body {
    margin-bottom: $m-size;
    @media (min-width: $desktop-breakpoint) {
        margin-bottom: $l-size;
    }
}

.list-item--message {
    align-items: center;
    color: $grey;
    justify-content: center;
    padding: $m-size;
    &:hover {
        background: none;
    }
}

.list-item__title {
    margin: 0;
    word-break: break-all;
}

.list-item__date {
    color: $grey;
    font-size: $font-size-small;
}

.list-item__data {
    margin: $s-size 0 0 0;
    @media (min-width: $desktop-breakpoint) {
        margin: 0;
        padding-left: $s-size;
    }
}