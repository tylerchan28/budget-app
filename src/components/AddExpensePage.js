import React from "react";
import { connect } from "react-redux"; // need this to call dispatch()
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.startAddExpense(expense);
        // dispatched from mapDispatchToProps (must use props.)
        // same functionality as props.dispatch(addExpense(expense)
        this.props.history.push("/"); //switches to dashboard after onSubmit w/ browser routing
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">   
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        onSubmit={this.onSubmit}
                    /> 
                </div>
            </div> // passes onSubmit prop to ExpenseForm
        )
    }
}



const mapDispatchToProps = (dispatch) => ({
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    })

export default connect(undefined, mapDispatchToProps)(AddExpensePage);

// mapDispatchToProps makes a more testable component
    // functions automatically dispatched when called
    // functions are passed as props to the component