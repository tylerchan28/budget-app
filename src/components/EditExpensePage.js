import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses"; 

const EditExpensePage = (props) => {
    return (
        <div>
            <ExpenseForm 
                expense={props.expense} // populates form when editing, different from expense variables below
                onSubmit={(expense) => { 
                    //dispatch action to edit expense
                    //redirect to dashboard
                    props.dispatch(editExpense(props.expense.id, expense)) // since expense={props.expense}
                    //  as the prop.expense changes, expense acts as the updater
                    props.history.push("/");
                    console.log("updated", expense);
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpense({ id: props.expense.id }));
                props.history.push("/");

            }}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)  
    }
}


export default connect(mapStateToProps)(EditExpensePage);

// set up mapStateToProps when you need more than dispatch()