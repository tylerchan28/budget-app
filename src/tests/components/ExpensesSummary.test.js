import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from '../../components/ExpensesSummary';

test("should correctly render ExpensesSummary with one expense", () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={250} />);
    expect(wrapper).toMatchSnapshot()
})

test("should correctly render ExpensesSummary with multiple expenses", () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={750} />);
    expect(wrapper).toMatchSnapshot()
})