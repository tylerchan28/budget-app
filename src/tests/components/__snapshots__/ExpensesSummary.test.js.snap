// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      You have 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $7.50
      </span>
      .
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      You have 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.50
      </span>
      .
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
